### PR TITLE
Implementation for VM creation on cluster

### PR DESCRIPTION
**Pull request by sireeshapotnuru**
This Implementation supports  VM creation on a cluster environment.

### CODE REVIEW

**Comment by sireeshapotnuru** [comment]
Need to Pass -TiPCluster "True" when invoking Run-LisaV2.ps1

File: Libraries/Framework-Azure.psm1 [hunk 1]
```diff
@@ -102,6 +102,16 @@ function Validate-Parameters {
     } else {
         LogMsg "Test parameters have been validated successfully. Continue running the test."
     }
+    if ($TiPCluster -eq "True") {
```
[the remark this answers is not in the document]
Comment (chvalean) [reply Libraries/Framework-Azure.psm1]: Probably defining the VM to be clustered would go as a VM flavor in the xml list of VM definitions?
Not sure to recommend the best option to do this.

File: Libraries/Framework-HyperV.psm1 [hunk 2]
```diff
@@ -589,6 +589,10 @@ function Run-Test {
     $currentTestResult.TestResult = GetFinalResultHeader -resultarr $resultArr
     if ($DeployVMPerEachTest -or $ExecuteTeardown) {
         LogMsg "VM CLEANUP ~~~~~~~~~~~~~~~~~~~~~~~"
+        foreach ($VM in $AllVMData) {
+            $HyperVGroup=$VM.HyperVGroupName
+            Add-VMGroupMember -Name "$HyperVGroup" -VM (Get-VM -name $VM.RoleName) -ComputerName $VM.HyperVHost
```
Comment (LiliDeng) [inline Libraries/Framework-HyperV.psm1]: Get-VM  add -ComputerName parameter?

File: Libraries/Framework-Azure.psm1 [hunk 3]
```diff
@@ -102,6 +102,16 @@ function Validate-Parameters {
...
+        Get-Cluster | out-null
```
Comment (LiliDeng) [inline Libraries/Framework-Azure.psm1]: Is that possible you run cluster testing remotely? if yes, should add computername parameter?

File: Libraries/Framework-Azure.psm1 (hunk 1 above)
Comment (LiliDeng) [inline Libraries/Framework-Azure.psm1]: You can make this parameter as [switch] type like EnableAcceleratedNetworking, then if ($TiPCluster) is enough to do judgement

File: Libraries/HyperV.psm1 [hunk 4]
```diff
@@ -477,6 +520,13 @@ Function CreateHyperVGroupDeployment([string]$HyperVGroup, $HyperVGroupNameXML,
                 $ErrorCount += 1
             }
         }
+        if($TiPCluster -eq "True") {
+            Move-VMStorage $CurrentVMName -DestinationStoragePath $DestinationOsVHDPath
```
[the remark this answers is not in the document]
Comment (sireeshapotnuru) [reply Libraries/HyperV.psm1]: Thanks Chris

File: Libraries/Framework-Azure.psm1 [hunk 5]
```diff
@@ -102,6 +102,16 @@ function Validate-Parameters {
...
+        if ($? -eq $False){
+            throw "Server doesn't have a cluster set up"
+        }
+        $clusterDir = Get-ClusterSharedVolume
+        if( -not $clusterDir) {
+            throw "No Cluster Shared Volume"
```
[the remark this answers is not in the document]
Comment (sireeshapotnuru) [reply Libraries/Framework-Azure.psm1]: modified throw message

File: Libraries/Framework-HyperV.psm1 (hunk 2 above)
Comment (sireeshapotnuru) [reply Libraries/Framework-HyperV.psm1]: added -ComputerName parameter

File: Libraries/Framework-Azure.psm1 (hunk 3 above)
Comment (sireeshapotnuru) [reply Libraries/Framework-Azure.psm1]: cant add computername  param to get-cluster command

File: Libraries/Framework-Azure.psm1 (hunk 1 above)
Comment (sireeshapotnuru) [reply Libraries/Framework-Azure.psm1]: Used new switch parameter -ClusteredVM

**Comment by sireeshapotnuru** [comment]
Now, we need to pass -ClusteredVM when invoking Run-LisaV2.ps1

File: AutomationManager.ps1 [hunk 6]
```diff
@@ -90,6 +91,10 @@ if($EnableAcceleratedNetworking) {
 	Set-Variable -Name EnableAcceleratedNetworking -Value $true -Scope Global
 }
 
+if($ClusteredVM) {
+	Set-Variable -Name ClusteredVM -Value $true -Scope Global
```
Comment (chvalean) [inline AutomationManager.ps1]: Simon recommend that we should not introduce any new global variables.
Please look into that, or if the effort is too much maybe @juhlee-microsoft  can confirm an exception.

**Comment by ader1990** [comment]
Hello,

This implementation seems pretty simple and might not catch all the issues when a live migration is tested. For this change it is mandatory to add an example test case, so that we can discuss on how this change can be better fitted to the needs of a real test case.

What are the requirements of a test case run in LISAV2 with a clustered VM?

As I can see, there cannot be a LISAV2 run with test cases that have clustered or non clustered VMs.

For start, can you add a test case that uses this framework change?

Thank you,
Adrian

**Comment by ader1990** [comment]
@sireeshapotnuru can you add the SECUREBOOT-MIGRATEVM code in the same PR?

File: AutomationManager.ps1 (hunk 6 above)
Comment (ader1990) [reply AutomationManager.ps1]: I think this clusteredvm should be a setuptype variable

File: Libraries/HyperV.psm1 [hunk 7]
```diff
@@ -67,6 +67,12 @@ Function DeployHyperVGroups ($xmlConfig, $setupType, $Distro, $getLogsIfFailed =
                         $retValue = $NULL
                     }
                 }
+                if($ClusteredVM) {
```
[the remark this answers is not in the document]
Comment (sireeshapotnuru) [reply Libraries/HyperV.psm1]: There is a lot of dependency for the vm to be in vmgroup, functions StartHyperVGroupVMs, GetAllHyperVDeployementData, uses HyperVGroupName to start vm & for setting $allVMData

**Comment by sireeshapotnuru** [comment]
@ader1990, I have added the Testcase "SECUREBOOT-MIGRATEVM" in this PR

File: AutomationManager.ps1 (hunk 6 above)
Comment (sireeshapotnuru) [reply AutomationManager.ps1]: @ader1990 do you want me need to add a new setuptype <ClusteredVM> with <isClusteredVM>True</isClusteredVM> and check using value of 
$xmlConfig.config.HyperV.Deployment.$setupType.isClusteredVM everywhere?

**Comment by sireeshapotnuru** [comment]
Added new setuptype "ClusteredVM"

File: Libraries/HyperV.psm1 [hunk 8]
```diff
@@ -114,7 +120,12 @@ Function CreateAllHyperVGroupDeployments($setupType, $xmlConfig, $Distro, $Debug
             }
 
             $SourceOsVHDPath = $xmlConfig.config.HyperV.Hosts.ChildNodes[$index].SourceOsVHDPath
-            $DestinationOsVHDPath = $xmlConfig.config.HyperV.Hosts.ChildNodes[$index].DestinationOsVHDPath
+            if ($setupType -eq "ClusteredVM") {
+                $clusterDir = Get-ClusterSharedVolume
```
[the remark this answers is not in the document]
Comment (ader1990) [reply Libraries/HyperV.psm1]: in this case, Run-LisaV2 base path is not used, this has to be documented

File: Libraries/HyperV.psm1 [hunk 9]
```diff
@@ -281,6 +292,35 @@ Function DeleteHyperVGroup([string]$HyperVGroupName, [string]$HyperVHost) {
         }
         Remove-VM -Name $vm.Name -ComputerName $HyperVHost -Force
         LogMsg "Hyper-V VM $($vm.Name) removed!"
+        if ($CurrentTestData.setupType -eq "ClusteredVM") {
+            LogMsg "Deleting VM on Cluster"
+            Get-Command "Get-ClusterResource" -ErrorAction SilentlyContinue
+            if ($?) {
+                $group = Get-ClusterGroup -ErrorAction SilentlyContinue
+                $vm_matches = $group -match $vm.Name
```
Comment (ader1990) [inline Libraries/HyperV.psm1]: be consistent, don t use bash like camel_case naming code in PowerShell

File: Libraries/HyperV.psm1 [hunk 10]
```diff
@@ -336,7 +376,12 @@ Function CreateHyperVGroupDeployment([string]$HyperVGroup, $HyperVGroupNameXML,
                 $hostNumber = $HyperVGroupXML.VirtualMachine.indexOf($VirtualMachine)
                 $HyperVHost = $xmlConfig.config.HyperV.Hosts.ChildNodes[$hostNumber].ServerName
                 $SourceOsVHDPath = $xmlConfig.config.HyperV.Hosts.ChildNodes[$hostNumber].SourceOsVHDPath
-                $DestinationOsVHDPath = $xmlConfig.config.HyperV.Hosts.ChildNodes[$hostNumber].DestinationOsVHDPath
+                if ($CurrentTestData.setupType -eq "ClusteredVM") {
```
Comment (ader1990) [inline Libraries/HyperV.psm1]: ditto

File: Libraries/HyperV.psm1 [hunk 11]
```diff
@@ -402,6 +447,7 @@ Function CreateHyperVGroupDeployment([string]$HyperVGroup, $HyperVGroupNameXML,
                 LogMsg "New-VM -Name $CurrentVMName -MemoryStartupBytes $CurrentVMMemory -BootDevice VHD -VHDPath $CurrentVMOsVHDPath -Generation $VMGeneration -Switch $($VMSwitches.Name) -ComputerName $HyperVHost"
                 $NewVM = New-VM -Name $CurrentVMName -MemoryStartupBytes $CurrentVMMemory -BootDevice VHD `
                     -VHDPath $CurrentVMOsVHDPath -Generation $VMGeneration -Switch $($VMSwitches.Name) -ComputerName $HyperVHost
+                Enable-VMIntegrationService -Name "Guest Service Interface" -vmName $CurrentVMName -ComputerName $HyperVHost
```
Comment (ader1990) [inline Libraries/HyperV.psm1]: why is this needed, all the integration services should have been already enabled

File: Libraries/HyperV.psm1 [hunk 12]
```diff
@@ -461,6 +507,13 @@ Function CreateHyperVGroupDeployment([string]$HyperVGroup, $HyperVGroupNameXML,
                 LogMsg "Prerequiste: Prepare OS Disk $CurrentVMOsVHDPath - Failed."
                 $ErrorCount += 1
             }
+            if($CurrentTestData.setupType -eq "ClusteredVM") {
+                Move-VMStorage $CurrentVMName -DestinationStoragePath $DestinationOsVHDPath
+                Add-ClusterVirtualMachineRole -VirtualMachine $CurrentVMName
+                if ($? -eq $False) {
+                    throw "High Availability configure for ${CurrentVMName} failed. The VM was not created in cluster"
```
Comment (ader1990) [inline Libraries/HyperV.psm1]: throw command does not work in this scenario, revisit how the code works above and use the $ErrorCount

File: XML/TestCases/FunctionalTests.xml [hunk 13]
```diff
@@ -2785,4 +2785,18 @@
         <Area>STORAGE</Area>
         <Tags>storage</Tags>
     </test>
+    <test>
+        <testName>SECUREBOOT-MIGRATEVM</testName>
+        <setupScript>.\TestScripts\Windows\SET-SECUREBOOT.ps1</setupScript>
+        <testScript>SECUREBOOT.ps1</testScript>
+        <Timeout>900</Timeout>
+        <setupType>ClusteredVM</setupType>
```
Comment (ader1990) [inline XML/TestCases/FunctionalTests.xml]: we were discussing to put the clusteredvm setuptype to the vm setuptype, and not the test case one.
is there a reason you put it here?

Comment (sireeshapotnuru) [reply XML/TestCases/FunctionalTests.xml]: we have introduced a new vm setuptype as "ClusteredVM" , which can be used for only cluster migration testcases and added the setuptype "ClusteredVM" in FunctionalTestsConfigurations.xml

File: Libraries/Framework-HyperV.psm1 [hunk 14]
```diff
@@ -581,6 +581,12 @@ function Run-Test {
 
     $currentTestResult.TestResult = GetFinalResultHeader -resultarr $resultArr
     LogMsg "VM CLEANUP ~~~~~~~~~~~~~~~~~~~~~~~"
+    if($CurrentTestData.setupType -eq "ClusteredVM") {
```
[the remark this answers is not in the document]
Comment (sireeshapotnuru) [reply Libraries/Framework-HyperV.psm1]: changed

File: Libraries/Framework-HyperV.psm1 [hunk 15]
```diff
@@ -581,6 +581,12 @@ function Run-Test {
...
+        foreach ($VM in $AllVMData) {
+            $HyperVGroup=$VM.HyperVGroupName
```
[the remark this answers is not in the document]
Comment (sireeshapotnuru) [reply Libraries/Framework-HyperV.psm1]: removed this variable and used $VM.HyperVGroupName

File: Libraries/HyperV.psm1 [hunk 16]
```diff
@@ -60,6 +60,12 @@ Function DeployHyperVGroups ($xmlConfig, $setupType, $Distro, $getLogsIfFailed =
                     $retValue = $NULL
                 }
             }
+            if($setupType -eq "ClusteredVM") {
```
[the remark this answers is not in the document]
Comment (sireeshapotnuru) [reply Libraries/HyperV.psm1]: broken into two lines

File: Libraries/HyperV.psm1 (hunk 8 above)
[the remark this answers is not in the document]
Comment (sireeshapotnuru) [reply Libraries/HyperV.psm1]: renamed to ClusterVolume

File: Libraries/HyperV.psm1 (hunk 11 above)
Comment (sireeshapotnuru) [reply Libraries/HyperV.psm1]: removed this line

File: Libraries/HyperV.psm1 (hunk 9 above)
Comment (sireeshapotnuru) [reply Libraries/HyperV.psm1]: modified variable name to MatchedVM

File: Libraries/Framework-HyperV.psm1 [hunk 17]
```diff
@@ -581,6 +581,12 @@ function Run-Test {
...
+            Add-VMGroupMember -Name "$HyperVGroup" -VM (Get-VM -name $VM.RoleName -ComputerName $VM.HyperVHost) -ComputerName $VM.HyperVHost
```
[the remark this answers is not in the document]
Comment (sireeshapotnuru) [reply Libraries/Framework-HyperV.psm1]: broken into two lines

File: Libraries/HyperV.psm1 (hunk 10 above)
Comment (sireeshapotnuru) [reply Libraries/HyperV.psm1]: changed

**Comment by sireeshapotnuru** [comment]
Latest Test Result:

[LISAv2 Test Results Summary]
Test Run On           : 11/22/2018 10:22:40
VHD Under Test        : SecbootGen2.vhdx
Total Test Cases      : 1 (1 Pass, 0 Fail, 0 Abort)
Total Time (dd:hh:mm) : 0:0:4
XML File              : TestConfiguration

   ID TestCaseName                                                 TestResult TestDuration(in minutes)
------------------------------------------------------------------------------------------------------
    1 SECUREBOOT-MIGRATEVM                                               PASS                 4.27

File: XML/VMConfigurations/FunctionalTestsConfigurations.xml [hunk 18]
```diff
@@ -682,4 +682,22 @@
             </VirtualMachine>
         </ResourceGroup>
     </ThreeVMsDM>
+    <ClusteredVM>
```
Comment (adriansuhov) [inline XML/VMConfigurations/FunctionalTestsConfigurations.xml]: Can this "ClusteredVM" tag be inside the SetupType? Let's take this scenario: If in the future we'll have a test case that requires 2 clustered VMs, then the framework should be changed. That means that we're stuck with only this setupType for a clustered VM(s) scenario. I would like to see a tag inside the Setuptype => this way we can reuse the clustered VMs logic to any SetupType.

Comment (sireeshapotnuru) [reply XML/VMConfigurations/FunctionalTestsConfigurations.xml]: As suggested added new setup "ClusteredSingleVM" with tag `<ClusteredVM>True</ClusteredVM>` and checking with $xmlConfig.config.HyperV.Deployment.($CurrentTestData.setupType).ClusteredVM

File: Libraries/Framework-HyperV.psm1 [hunk 19]
```diff
@@ -581,6 +581,12 @@ function Run-Test {
 
     $currentTestResult.TestResult = GetFinalResultHeader -resultarr $resultArr
     LogMsg "VM CLEANUP ~~~~~~~~~~~~~~~~~~~~~~~"
+    if ($xmlConfig.config.HyperV.Deployment.($CurrentTestData.setupType).ClusteredVM) {
+        foreach ($VM in $AllVMData) {
+            Add-VMGroupMember -Name $VM.HyperVGroupName -VM (Get-VM -name $VM.RoleName -ComputerName $VM.HyperVHost) `
+             -ComputerName $VM.HyperVHost
```
Comment (ader1990) [inline Libraries/Framework-HyperV.psm1]: leave a tab or 4 spaces for the indent

**Comment by ader1990** [comment]
> * to another node irrespective of migration type
> * Should be able to stop the Cluster service on a node in

Hello,

I was referring to lisav2 running test cases using clustered vms.
I have some questions:
1. What are the requirements of the environment in addition to the hyperv lisav2 requirements?
2. Can I run other test cases in the same lisav2 run?

Please answer to all these questions in the commit message too.

File: Libraries/Framework-HyperV.psm1 [hunk 20]
```diff
@@ -581,6 +581,12 @@ function Run-Test {
 
     $currentTestResult.TestResult = GetFinalResultHeader -resultarr $resultArr
     LogMsg "VM CLEANUP ~~~~~~~~~~~~~~~~~~~~~~~"
+    if ($xmlConfig.config.HyperV.Deployment.($CurrentTestData.setupType).ClusteredVM) {
```
Comment (ader1990) [inline Libraries/Framework-HyperV.psm1]: what do you mean $currenttestdata.setuptype?

Comment (ader1990) [reply Libraries/Framework-HyperV.psm1]: $CurrentTestData.setupType is the actual VM setup type?

File: Libraries/HyperV.psm1 [hunk 21]
```diff
@@ -281,6 +291,35 @@ Function DeleteHyperVGroup([string]$HyperVGroupName, [string]$HyperVHost) {
...
+                $MatchedVM = $group -match $vm.Name
+                foreach ($vm in $MatchedVM) {
+                    Remove-ClusterGroup -Name $vm.name -RemoveResources -Force
+                    if (-not $?) {
+                        LogErr "Failed to remove Cluster Role for VM $vm"
+                        return $False
+                    }
+                    LogMsg "Cleanup was successful for $vm on Cluster"
+                }
+                # Also remove VM from second node if it's located there
```
Comment (ader1990) [inline Libraries/HyperV.psm1]: don t rely on the fact the there is only one other cluster node, remove the vm from all nodes.
this is a framework change and changes here should be generic, even if the test cases rely on the fact that there is a setup with only two cluster nodes

File: Libraries/HyperV.psm1 [hunk 22]
```diff
@@ -461,6 +505,14 @@ Function CreateHyperVGroupDeployment([string]$HyperVGroup, $HyperVGroupNameXML,
                 LogMsg "Prerequiste: Prepare OS Disk $CurrentVMOsVHDPath - Failed."
                 $ErrorCount += 1
             }
+            if ($xmlConfig.config.HyperV.Deployment.($CurrentTestData.setupType).ClusteredVM) {
+                Move-VMStorage $CurrentVMName -DestinationStoragePath $DestinationOsVHDPath
+                Add-ClusterVirtualMachineRole -VirtualMachine $CurrentVMName
+                if ($? -eq $False) {
+                    LogErr "High Availability configure for ${CurrentVMName} failed. The VM was not created in cluster"
```
Comment (ader1990) [inline Libraries/HyperV.psm1]: put something like
"VM ${CurrentVMName} could not be added to the Hyper-V cluster."

File: Libraries/Framework-HyperV.psm1 (hunk 20 above)
Comment (sireeshapotnuru) [reply Libraries/Framework-HyperV.psm1]: Yes

**Comment by sireeshapotnuru** [comment]
> > * to another node irrespective of migration type
> > * Should be able to stop the Cluster service on a node in
> 
> Hello,
> 
> I was referring to lisav2 running test cases using clustered vms.
> I have some questions:
> 
> 1. What are the requirements of the environment in addition to the hyperv lisav2 requirements?
> 2. Can I run other test cases in the same lisav2 run?
> 
> Please answer to all these questions in the commit message too.

Answers as below: 
What are the requirements of the environment in addition to the hyperv lisav2 requirements?
	1) Should have a Cluster setup with atleast two nodes and one storage volume

Can I run other test cases in the same lisav2 run?
	2) yes we can run both Non-clustered and Clustered testcases in same run

Updated the same in commit message too

File: Libraries/Framework-HyperV.psm1 (hunk 19 above)
Comment (sireeshapotnuru) [reply Libraries/Framework-HyperV.psm1]: added tab

File: Libraries/HyperV.psm1 (hunk 22 above)
Comment (sireeshapotnuru) [reply Libraries/HyperV.psm1]: modified logerr line

File: Libraries/HyperV.psm1 (hunk 21 above)
Comment (sireeshapotnuru) [reply Libraries/HyperV.psm1]: changed logic which checks vm on other nodes and removes it
